presto-spi/src/main/java/io/prestosql/spi/tracer/DefaultTracer.java

```diff
@@ -0,0 +1,297 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.spi.tracer;
+
+import io.prestosql.spi.eventlistener.TracerEvent;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+
+import static java.util.Objects.requireNonNull;
+
+// Tracer keeps track of Presto execution unit contexts and sends debugging purpose events via its emitter.
+// Users should not build complex metrics based on the action events.
+public class DefaultTracer
+        implements Tracer
+{
+    private final boolean enabled;
+    private final Consumer<TracerEvent> emitter;
+    private final String queryId;
+    private final Optional<String> stageId;
+    private final Optional<String> taskId;
+    private final Optional<String> pipelineId;
+    private final Optional<String> splitId;
+    private final Optional<String> operatorId;
+    // presto worker node emitting tracer events
+    private final Optional<String> nodeId;
+    private final Optional<String> host;
+    private final Optional<Integer> port;
+
+    private DefaultTracer(
+            boolean isEnabled,
+            Consumer<TracerEvent> emitter,
+            String queryId,
+            Optional<String> stageId,
+            Optional<String> taskId,
+            Optional<String> pipelineId,
+            Optional<String> splitId,
+            Optional<String> operatorId,
+            Optional<String> nodeId,
+            Optional<String> host,
```

Comment:
let's combine host and port in the same string, this information is not heirarchical